File: graphql/udemy1-graphql17/main/GraphQL-Directives/src/main/java/com/accounts/service/BankService.java
package com.accounts.service;

import com.accounts.entity.BankAccount;
import com.accounts.domain.Client;

import com.accounts.exceptions.AccountNotFoundException;
import com.accounts.exceptions.ClientNotFoundException;
import com.accounts.repo.BankAccountRepo;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
@Slf4j
public class BankService {

    private static List<Client> clients = Arrays.asList(
            new Client(100L,  "John", "T.", "Doe"),
            new Client(101L, "Emma", "B.", "Smith"),
            new Client(102L, "James", "R.", "Brown"),
            new Client(103L, "Olivia", "S.", "Johnson"),
            new Client(104L, "William", "K.", "Jones")
    );
    @Autowired
    BankAccountRepo repo;

    public void save(BankAccount account) {
        if (validClient(account))
            repo.save(account);
        else
            throw new ClientNotFoundException("Client Not Found");
    }

    public BankAccount modify(BankAccount account) {
        if (validClient(account))
            repo.save(account);
        else
            throw new ClientNotFoundException("Client Not Found");

        return account;
    }

    public List<BankAccount> getAccounts() {
        return repo.findAll();
    }

    public BankAccount accountById(Long accountId) {
        if (repo.findById(accountId).isPresent()) {
            return repo.findById(accountId).get();
        }
        throw new AccountNotFoundException("Account Not Found");
    }

    public Boolean delete(Long accountId) {
        if (repo.findById(accountId).isPresent()){
            repo.delete(repo.findById(accountId).get());
            return true;
        }
        return false;
    }

    private List<Client> getClients () {
        return clients;
    }

    public Map<BankAccount, Client> getBankAccountClientMap(List<BankAccount> bankAccounts) {
        // Collect all client IDs from the list of bank accounts
        Set<Long> clientIds = bankAccounts.stream()
                .map(BankAccount::getClientId)
                .collect(Collectors.toSet());

        // Fetch client for all collected IDs
        List<Client> clients = getClients ().stream()
                .filter(client -> clientIds.contains(client.getId()))
                .collect(Collectors.toList());

        // Map each bank account to its corresponding client
        return clients.stream()
                .collect(Collectors.toMap(
                        client -> bankAccounts.stream()
                                .filter(bankAccount -> bankAccount.getClientId().equals(client.getId()))
                                .findFirst()
                                .orElse(null),
                        client -> client
                ));
    }

    private boolean validClient(BankAccount account) {
        return getClients ().stream()
                .filter(client -> client.getId().equals(account.getClientId())).findAny().isPresent();
    }
}

File: graphql/graphqlex/ex1/src/main/kotlin/com/yscorp/ex1/BankService.kt
package com.yscorp.ex1

import com.yscorp.ex1.domain.Client
import com.yscorp.ex1.entity.BankAccount
import com.yscorp.ex1.exception.ClientNotFoundException
import com.yscorp.ex1.repository.BankAccountRepo
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.stereotype.Service
import javax.security.auth.login.AccountNotFoundException

@Service
class BankService(
    private val repo: BankAccountRepo,
) {


    private val log = KotlinLogging.logger {}

    var clients: List<Client> = listOf(
        Client(100L, "1", "John", "T.", "Doe"),
        Client(101L, "2", "Emma", "B.", "Smith"),
        Client(102L, "3", "James", "R.", "Brown"),
        Client(103L, "4", "Olivia", "S.", "Johnson"),
        Client(104L, "5", "William", "K.", "Jones")
    )


    fun save(account: BankAccount) {
        if (validClient(account)) repo.save(account)
        else throw ClientNotFoundException("Client Not Found")
    }

    fun modify(account: BankAccount): BankAccount {
        if (validClient(account)) repo.save(account)
        else throw ClientNotFoundException("Client Not Found")

        return account
    }

    fun getAccounts(): List<BankAccount> {
        return repo.findAll()
    }

    fun accountById(accountId: Long): BankAccount {
        if (repo.findById(accountId).isPresent) {
            return repo.findById(accountId).get()
        }
        throw AccountNotFoundException("Account Not Found")
    }

    fun delete(accountId: Long): Boolean {
        if (repo.findById(accountId).isPresent) {
            repo.delete(repo.findById(accountId).get())
            return true
        }
        return false
    }

    private fun getClients(): List<Client> {
        return clients
    }

    fun getBankAccountClientMap(bankAccounts: List<BankAccount>): Map<BankAccount, Client> {
        // Collect all client IDs from the list of bank accounts
        val clientIds = bankAccounts
            .map { it.clientId }

        // Fetch client for all collected IDs
        val clients = getClients()
            .filter { client: Client -> clientIds.contains(client.id) }

        // Map each bank account to its corresponding client
        return clients.associateBy { client ->
            bankAccounts.firstOrNull { bankAccount ->
                bankAccount.clientId == client.id
            }!!
        }

    }

    private fun validClient(account: BankAccount): Boolean {
        return getClients()
            .firstOrNull { client: Client -> client.id == account.clientId } != null
    }
}

